osd: Relax missing entry assert for partial writes. (#65770)
ceph/ceph: doc/dev/osd_internals/erasure_coding/direct_reads.rst

===========================
Erasure Coding Direct Reads
===========================

This document covers the design for EC Direct Reads, it also covers split 
reads for large replica IOs.

Basic design
------------

Small reads
~~~~~~~~~~~

> Currently, all reads in an erasure-coded pool are directed to the primary

Reviewer: s/in/to/

Author: The dev doc was merged by mistake.  I will remove

which is implemented in the user-space ``librados`` library, must be replicated
within the kernel drivers.

This involves implementing the ``SplitRead`` mechanism, including the
calculation of target OSDs, the parallel submission of I/O requests, and the
subsequent reassembly (concatenation or de-striping) of data. This is a
non-trivial task, as kernel development brings additional complexity regarding
memory management, asynchronous I/O handling, and ensuring system stability.
The design and implementation of these kernel-level changes are recognised as
a necessary work item, but the specific approach will require a separate,
detailed investigation. Additionally, the challenge of submitting code to the
kernel itself may prove an interesting challenge!

Performance, QoS and throttling
-------------------------------

The direct read mechanism fundamentally alters the I/O pattern from the
perspective of the OSDs. Where previously a client read resulted in a single
I/O operation to the primary OSD, this design will now submit multiple, smaller
I/O operations in parallel to different OSDs.

The balance of network operations will change and this will need to be carefully

> assesed in performance evaluation, so that we can provide sufficient data to the

Reviewer 2: I think this should be "assessed" and not "assesed".

```suggestion
assessed in performance evaluation, so that we can provide sufficient data to the
```

> end user, to help make their decision about enabling this optional feature.

Reviewer 2: Should this comma be omitted?

```suggestion
end user to help make their decision about enabling this optional feature.
```

> The splitting will significantly reduce the network traffic withing the cluster.

Reviewer 2: Should be "within" and not "withing".

```suggestion
The splitting will significantly reduce the network traffic within the cluster.
```

The network load on the public, client network will see a more complex change:

- Small IOs (sub-chunk) will see no increase in traffic at all.
- Very large IOs will see an increase in the number of messages, but this is

>   likely insignificant compared to the associated data transfers

Reviewer 2: Missing full stop.

```suggestion
  likely insignificant compared to the associated data transfers.
```

- For the intermediate IOs, the performance considerations are more nuanced, the
  splitting up of IOs will help to further distribute network traffic, but will
  lead to a larger overhead, as the IO count will increase.

CPU Load
~~~~~~~~

As part of the implementation, the impact on CPU utilization on the client will
be measured. The "destripe" mechanism is CPU intensive and results in
non-contiguous buffers.  This may be significant for some applications where

> the ceph client is completing for CPU resources with the client application.

Reviewer 2: Capitalize Ceph?

```suggestion
the Ceph client is completing for CPU resources with the client application.
```

may change. The intent is to support ops such that the vast majority of IO in
realistic work loads of RBD, CephFS and RGW perform EC direct reads. At the
time of writing, the following seems like a reasonable limitation:

- Objecter ops must contain a single op
- Read
- Sparse read

It is possible, but more complex to permit:

- Multiple reads/sparse reads
- Attribute reads (when performed with a read or sparse reads)

Sparse Reads
~~~~~~~~~~~~

Sparse reads are currently treated as full reads by EC. Direct reads will
support sparse reads. A new synchronous code path through the OSD EC Backend is
required and implementing this such that it supports sparse reads is simple
enough to be worth implementing

No attempt will be made to support sparse reads on any other EC reads. This
means that in failure scenarios, the sparseness of the backend will disappear.
This would likely be a significant issue for encryption, so will need to be
addressed before encryption is supported. No client that supports EC can have a
dependency on sparse reads being fully implemented, so this should not cause
any regressions in the clients.

Object (user) Version
---------------------

> Clients can (and regularly do) request an “object version”. In the OSD this is
> known as the “user version” and is store in the OI along with the OSD’s

Reviewer 2: Two pairs of Unicode quotation marks and apostrophe which might be better as ASCII.

Also, "is stored" instead of "is store".

```suggestion
Clients can (and regularly do) request an "object version". In the OSD this is
known as the "user version" and is stored in the OI along with the OSD's
```

Already covered by IO exerciser, currently sends all IOs through librados which
calls Objecter. Running through this should give us coverage of various reads.

Stale Maps/Fallback path
++++++++++++++++++++++++

IO Exerciser has two injects for reads currently:

#. Type = 0 injects EIO when reading data, shard indicates which shard to fail.
   The inject is sent to the primary OSD
#. Type = 1 makes reads pretend a shard is missing which will immediately
   trigger extra reads for a decode, shard indicates which shard to fail. The
   inject is sent to the primary OSD

As it stands, these will not work with Direct Reads and we will need to make
modifications to both the IO Exerciser so they send to the shard they want to
inject, along with modifications to the injects so they can be called on the
synchronous read path.

Missing Objects
+++++++++++++++

We could add a Type = 2 inject similar to Type = 0 where instead of returning
EIO it returns our new error code. We would need to either have a hook into
objecter to tell it to start sending direct reads again in the same epoch, or
force a new epoch following this. This would allow us to exercise the fallback
path of the client.

> Possibly we might want to enhance the —testrecovery flag to be something like
> —testrecovery=“read:2” or something to ensure we are able to use this new

Reviewer 2: More Unicode: quotation marks and an M dash. Also propose to put the switches inside double backticks to render them as preformatted/code.

```suggestion
Possibly we might want to enhance the ``--testrecovery`` flag to be something like
``--testrecovery="read:2"`` or something to ensure we are able to use this new
```
